runtime-sdk/src/crypto/signature/ed25519.rs

```diff
@@ -118,60 +118,180 @@ impl From<PublicKey> for CorePublicKey {
 
 /// A memory-backed signer for Ed25519.
 pub struct MemorySigner {
-    sk: ed25519_dalek::ExpandedSecretKey,
+    key: Key,
+}
+
+/// The original version of this signer returned the "expanded" secret key from `to_bytes`.
+/// If the contract passes in one of these, it's gets the same thing out. New invocations get the "proper" version.
+enum Key {
+    Expanded {
+        esk: ed25519_dalek::hazmat::ExpandedSecretKey,
+        /// The hash output that is used to create the "expanded" secret key. It is stored to return from `from_bytes` because `esk` is a function of the hash, not the hash directly.
+        hash: [u8; 64],
+    },
+    Regular(ed25519_dalek::SigningKey),
+}
+
+impl Key {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        if bytes.len() == 64 {
```

Comment:
I generally don't `match` on non-string literals to avoid the extra indentation and syntax noise, but I'll change it to a `match` since you prefer it.